### DISCUSSION TITLE
Add bookmarks sorting options

### DISCUSSION THREAD

**Discussion by PetbkA** [organicmaps/organicmaps#484] [Ideas]
**OS**: Android 10.0 (LineageOS 17.1)
**App version:** version 2021.05.08-7-g19ac5fb5-Google

Changing sorting is possible only inside of the list of bookmarks. It would be nice to have sorting options for the lists themselves.

Button for changing sorting can be in upper right corner of the root of `Bookmarks` menu, and sorting should be applied to the menu and to the "list of lists" in bookmark editor

**Comment by biodranik** [comment]
Which sorting modes do you need and what is your use case?

**Reply by palzkillm** [reply]
I‘d just need a simple Alphanumeric sort. Reasoning is, that my 150+ bookmark lists are currently completely unsorted.
![Uploading FA8FE676-B41E-4FD2-8235-B499EDF36C56.png…]()

This makes it very tedious to add new bookmarks to a specific list.
![9051A706-8387-403F-B5BE-72EB0E58A1E1](https://user-images.githubusercontent.com/107746723/183268608-a6e7acce-26ab-4481-8abf-31c9589a7ac6.png)

**Comment by PetbkA** [comment]
The ones that's for bookmarks (excluding Distance) is enough:
1. Default (as I guess, sorts by date added);
2. Name.

With sorting by name it's easier to update the lists that you've created a long time ago - you will find the list by alphabet faster

**Reply by Jeremboo** [reply]
Maybe it could be good as a start to see the bookmarks automatically sorted by name!

I use to name my folders like:
- "0_France", "1_Canada" 
- "20200102_TravelToParis", "20210410_Travel to Madrid"

That can be a quick first solution I guess. Otherwise sorting by date or name should be good too!

**Comment by Jeremboo** [comment]
Maybe instead of the sorting option you can have a feature to reorder the list you have. Like this:
![Aug-10-2021 13-20-30](https://user-images.githubusercontent.com/6682092/128858227-7008cdeb-e4ac-413c-a7fb-dc4eaf20cf8e.gif)

**Comment by Esmolante** [comment]
I agree. It would be nice to be able to sort favorites folders at will, or at least to sort them alphabetically, as well as their contents by name and distance from us (the same as POI or search tool).

**Comment by palzkillm** [comment]
Just upvoted. I have > 150 Bookmark lists, and working with them in an unsorted list is tedious.
Question to the Devs: Is there anything specific making this difficult to implement? Maybe there are ways to simplify implementation?

**Comment by dobbyt1** [comment]
Here is my vote of support for a review of the Bookmark system.

Love old MapsWithMe and now it's younger sibling so I'm here to give the a user's experience and help make Organic even better.

Bookmarks, be they physical in the pages of a book or virtual in a software folder they are intended to help us humans return to a place we have recorded as worth being again.

If your library of notes is a bit large and random to the point of chaotic, like a depressing pile papers tipped out of a refuse bin, you may as well not have the records.  How do find anything of interest?   Filing!

A bin is a kind of recepticle for notes but lacks a helpful retrieval system. Introduce an ordered filing scheme then we can conquer the world!

I do not need to tell you how useful a tiered filing system is. It is not sexy but it is necessary. 
Spain/Malaga/Restaurants....makes sense.  
(1400notes/314notes/10notes.)
So does:
Spain/Fuengirola/Accommodation.
(1400notes/257notes/7notes). It makes it so much more manageable.

At Group level, my vote, goes for the ability to order groups by Name and Date Created but I would also like to see a group within group system as illustrated above. 

Thanks to all involved for putting together a fine community made product. It can only get better.

Regards.

**Reply by biodranik** [reply]
@dobbyt1 thanks for the feedback! What do you think, will a convenient search help you? E.g. if you type the group/bookmark name in the main search field?

**Comment by dobbyt1** [comment]
Alexander,

Thank you for getting in touch and for giving me this chance to offer some
thoughts on the bookmark search feature in Organic Maps.

I delayed responding because I wanted to give it some thought and give you
a balanced view of what you are asking from real situations.

Organic Maps is an app that, at its simplest, is about geography and moving
around, about where we are and where we are going. And essentially the app
already filters out irrelevant information, in that, it locates us in a
single precise ten metre square - out of the millions of squares on this
planet.

A simple search of a bookmark, be it from the Main Search field, or a more
targeted Bookmarks Only field will have its issues.

Looking for a bookmark is going to require me to know precisely how I
recorded the note in the first place. If I spelt it wrong, or there are
legitimate variations, my required bookmark will not be among the results.
Try 'Tourist Office'.   I could have that bookmark recorded by me correctly
but I could have copied the entry from the OpenStreet note on the map for
my use later. The second bookmark, Oficina de Turismo, would be missed as
it is in the local language and therefore skipped.

Currently a search in the Main Search field does produce a multi-lingual
selection for the phrase Tourist Office but it appears there is a category
subtitle in the record of each of the entries.  The title of the
sub-category is in english and I suspect this is why this search is so
comprehensive and covers so many different languages.  Will Bookmarks need
sub-categories to filter in this way? The human is lazy, and most are not
librarians, and will not want to fill out any more fields than have too.

Many bookmarks are derived from the OpenStreet information where a feature
is found and a bookmark is created, on or nearby the original point to make
the bookmark even more useful to the user.  I have a bookmarked supermarket
in France that I logged from OSM building location info then modified my
pin's location to specify the entrance to the car park. I will be driving a
big vehicle so this modification helps with local navigation.  The original
note is titled Carrefour Market (supermarket).  To me the record is titled
'Carrefour, Thiex', the name of the shop and the town name, but it also
carries the record sub-category information which now reads (Parking
Aisle). The title works for me, the human, as i plan my journey and am
looking at a selection of pins on the selected area of the map but if I was
to search for it by a general search feature I would have to have memorised
the name of the supermarket or the town accurately in advance, out of
hundreds, as 'supermarket' would not find my bookmark.

Sometimes OSM is also vague.  An untitled restaurant is described as
'Building' on OSM. So the derived bookmark inherits that uneditable
subcategory and so it is no good for searching for 'restaurant'. So that
leaves me searching by title. "What was it called?", "Lay..." "Lee..." or
was it "Loy...".

Maybe there is a better way.

In the real world if I am looking for a fork and I rely on a ' Main Search'
type search I am likely to get a lot of irrelevant but valid answers
without more information or filtering.

Fork - Kitchen,  Fork - Bitcoin,  Fork - Garden,  Fork - Bicycle,  Fork -
Road,  Fork - Software system call.  All valid but maybe we need context to
improve matters.

I do not think a free text search facility is what is needed to improve
searching for bookmarks.

The fork I was looking for was in the garden so maybe it would have been
better if I had told the search feature that first. And even better let you
know the garden was in Oliver Road in Bury St. Edmunds. Not a well known
town so I had better let you know that it is in England.

So now, like a Library or a computer filing or labeling system I can find
the fork by a list within lists system; England; Bury St. Edmunds; Oliver
Road; number; Garden. And being human, an ordering system is important.  Bury
St. Edmunds needs to be in an alpha-numeric list to be found because of the
thousands of town names possible.

The Tourist Office I was looking for could be logged as Spain; Malaga;
'Oficina de Turismo'. (One of three)

The restaurant with the sub category (Building) could be logged as France;
'Relais Routiers; Claire de Lie'.

The supermarket Carrefour could be logged as France; Route South Sept 2022;
'Carrefour, Thiex'.

Each is easily and logically found by a human because each list and sublist
is shown in an alpha-numeric order.

If folders within folders are not possible would the Label system Google
use in Gmail be a possibility.

I hope this has been of use and provides food for thought.  If you have any
further questions I am happy to try and help.

Regards

Tony

**Reply by biodranik** [reply]
1. Our search can take into account spelling errors.
2. Object types/translations can also be taken into an account, we have a list of all translations in the search core.
3. You can specify a city/town and country/state name in the search to see the closest results first. The same can be done for bookmarks.
4. If you name your bookmark somehow, and then you are visually looking/searching that bookmark's name by scrolling the list of bookmarks, does it mean that you can try to type that name (or part of it) in the search and find the same bookmark faster?

Before we proceed with a complicated implementation of subfolders/labels, let's evaluate how many popular/important cases can be fixed in an easier/faster way for us and for users using the search.

**Comment by PetbkA** [comment]
Funny how this issue went to something bigger than adding sorting options. Guys asking to implement search and so on should've better create or join to another issue.

Wil try to be concise:
1. Adding search in bookmarks section isn't necessary 'cause it's better to just show bookmarks in search results. If someone wants to filter search by bookmarks specifically, some keyword like "bookmark" can be used (the problem is how to notify users about the feature);
1. Adding manual sorting is hard to implement, and I doubt that the average user will frequently need it;
1. As for differences in names, it's the user who knows better what name of the bookmark should be. I'm using english version of OM and pretty often change the name of the bookmark I'm adding 'cause it should be in russian, but bookmark gets english name due to OM's locale. Never had a problem with that and always could find the specific bookmark;
1. Adding subcategories/subfolders is very painful task, mostly in terms of export to KML which I doubt that has subfolders support. Import toother apps can raise problemstoo. BTW, is there other maps apps that have subfolders?
1. May be subfolders will ease pain for users that have hundreds of folders and thousands of bookmarks inside them, but the main question is WHY they need so many bookmarks? Usually you create many bookmarks when travelling → Usually your trip is time limited, and you just can't visit so many POIs → So the bookmarks for one trip will hardly require so many bookmarks → After the trip is over it's better to backup and remove it's bookmarks. I never had more than 10-20 folders and several hundreds of bookmarks for a trip, so never had a need to use subfolders

**Reply by palzkillm** [reply]
For me with > 150 Bookmarks, a simple A-Z sort has always been good enough.

**Reply by Esmolante** [reply]
1. OK
2. OK
3. OK
4. iGO 8, but OK.
5. When I travel for leisure, with the motorhome, I collect restaurants, places to sleep, cheap petrol stations, parking and water. I have hundreds of them all over Europe. OM shows them in order of incorporation, which is not a very useful criterion when you are driving. As I understand it, there should be two ways of sorting the list: alphabetically and by distance from our position. 
An example: if I'm driving and I need to look for a place to spend the night, I don't know beforehand the name I'm looking for, I just want to see if I have anything saved in that area. The same for a petrol station...

BTW: OM has a very good searching tool. It is one of its strengths. Congratulations. ;)

**Comment by dansseg** [comment]
I would also like to sort the bookmark lists manually by drag & drop. 

Is it maybe possible to sort the lists somehow else? (e.g. by editing a file on the file system?) I think I sorted them a long time ago somehow (or was this with maps.me?)

**Comment by Hoshi98** [comment]
a system obviously designed by neckbeards who lack the ability to understand how real people interact with software.
you're really going to have no sorting or subfolders for people who may have hundreds/thousands of poi's? lol.
get real, not going to spend hours scrolling through a mile long list of objects. deleted.

**Reply by biodranik** [reply]
@Hoshi98 did you use the latest app version? Bookmark sorting was implemented and merged in April: https://github.com/organicmaps/organicmaps/pull/7648
